### PR TITLE
Fix disabled 'eth_sign' event & add 'Request Disabled' event type

### LINKED ISSUE #17974: [Bug]: When `eth_sign` is disabled we fire a `Signature Approved` event everytime we call the RPC method
### Describe the bug

**Problem**: when `eth_sign` is disabled from Advanced settings, we fire a `Signature Approved` event everytime we call the RPC method. This should not be the case.

This issue was introduced in 10.25.0, by this PR https://github.com/MetaMask/metamask-extension/pull/17308

https://user-images.githubusercontent.com/54408225/222666667-07b91fba-798c-4daf-8f95-ab64391ba351.mp4

### Steps to reproduce

1. Open Segment or open background console for check the network requests to Segment
2. Disable the functionality of `eth_sign` from MetaMask Advanced Settings
3. On another window, go to test dapp
4. Connect to test dapp
5. Click Eth Sign
6. See how the `Signature Approved` event is fired

### Error messages or log output

_No response_

### Version

10.26.0

### Build type

None

### Browser

Chrome

### Operating system

Linux

### Hardware wallet

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by digiwand**
## Explanation

Fixes https://github.com/MetaMask/metamask-extension/issues/17974

- Fixes bug where 'Signature Approved' event was incorrectly called when 'eth_sign' rpc method is called while the method is disabled by advanced settings
- Replaces 'Signature Approved' event with new 'Signature Failed' event when the disabled 'eth_sign' event is called (spec discussed [here in Slack](https://consensys.slack.com/archives/G01FXS64DFA/p1678102135498189))

**note:** 
1. sometimes, the 'Signature Failed' effect may be received before 'Signature Request', but the 'sent_at' time still records it being sent earlier. 
2. clean-up code, including more use of constants, is applied to this follow-up PR https://github.com/MetaMask/metamask-extension/pull/18008

### After

Screen Shot 2023-03-13 at 12 10 39 PM
Screen Shot 2023-03-13 at 12 10 45 PM

## Manual Testing Steps

1. Open Segment or open background console for check the network requests to Segment
2. Disable the functionality of eth_sign from MetaMask Advanced Settings
3. On another window, go to test dapp
4. Connect to test dapp
5. Click Eth Sign
6. See Signature Request and Request Denied events are fired. Signature Approval should no longer be fired.

## Pre-merge author checklist

- [x] I've clearly explained:
  - [x] What problem this PR is solving
  - [x] How this problem was solved
  - [x] How reviewers can test my changes
- [x] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [x] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [x] PR is linked to the appropriate GitHub issue
- [x] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

### CODE REVIEW

**Comment by digiwand** [comment]
moving back to draft. More discussions are being had to plan how we will want to handle this event

File: app/scripts/lib/createRPCMethodTrackingMiddleware.js [hunk 1]
```diff
@@ -138,6 +140,12 @@ export default function createRPCMethodTrackingMiddleware({
     // keys for the various events in the flow.
     const eventType = EVENT_NAME_MAP[method];
 
+    const isDisabledEthSignAdvancedSetting =
+      method === MESSAGE_TYPE.ETH_SIGN &&
+      res.error?.code === errorCodes.rpc.methodNotFound;
+
+    const isDisabledRequest = isDisabledEthSignAdvancedSetting;
```
Comment (jpuri) [inline app/scripts/lib/createRPCMethodTrackingMiddleware.js:147]: hey, reason for duplicating the variable ?

**Review by jpuri — DISMISSED** [review]
Great work 👍

File: app/scripts/lib/createRPCMethodTrackingMiddleware.js (hunk 1 above)
Comment (digiwand) [reply app/scripts/lib/createRPCMethodTrackingMiddleware.js:147]: heyy, I did this for readability down the line

Comment (digiwand) [reply app/scripts/lib/createRPCMethodTrackingMiddleware.js:147]: hey, I did this for readability down the line. hope it helps, unless it's just me - In that case, happy to change it back

**Review by jpuri — DISMISSED** [review]

**Review by jpuri — DISMISSED** [review]

**Review by blackdevelopa — DISMISSED** [review]
LGTM

**Review by seaona — CHANGES_REQUESTED** [review]
oh I am still seeing the `Signature Requested` and `Signature Approved` event sent to Segment, with `eth_sign` disabled and no `Request Disabled` event thrown.

I believe I have the last changes :thinking: : could you double check this behaviour to discard it's just me? :pray: :

https://user-images.githubusercontent.com/54408225/224732586-4e35c8fe-63c5-4400-80ef-66a18b8e515b.mp4

**Comment by digiwand** [comment]
oof, great catch @seaona! 

Derp.. when we decided to exclude `Signature Request`, I updated the code and moved the `res.error?.code` check up, except the response, `res`, doesn't receive the values at this point... 

I'm reworking this PR now

**Review by jpuri — APPROVED** [review]

**Review by seaona — APPROVED** [review]

**Pull request merged**